File: src/TextFix/Core/HotkeyBinding.cs
// SPDX-License-Identifier: GPL-3.0-only
using TextFix.Interop;

namespace TextFix.Core;

/// <summary>
/// A global hotkey, stored in config.json as text ("Ctrl+Alt+J", "F13") so it stays hand-editable,
/// and converted here to the shapes RegisterHotKey and the msctls_hotkey32 control expect.
/// </summary>
internal readonly record struct HotkeyBinding(uint Modifiers, uint VirtualKey)
{
    // msctls_hotkey32 reports its own modifier bits, which are not the RegisterHotKey ones.
    private const int HotkeyfShift = 0x01;
    private const int HotkeyfControl = 0x02;
    private const int HotkeyfAlt = 0x04;

    internal bool IsEmpty => VirtualKey == 0;

    internal static readonly HotkeyBinding None = new(0, 0);

    internal static bool TryParse(string? text, out HotkeyBinding binding)
    {
        binding = None;
        if (string.IsNullOrWhiteSpace(text)) return false;

        uint modifiers = 0;
        uint vk = 0;
        foreach (string rawPart in text.Split('+', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            string part = rawPart.ToLowerInvariant();
            switch (part)
            {
                case "ctrl" or "control":
                    modifiers |= Win32.MOD_CONTROL;
                    continue;
                case "alt" or "menu":
                    modifiers |= Win32.MOD_ALT;
                    continue;
                case "shift":
                    modifiers |= Win32.MOD_SHIFT;
                    continue;
                case "win" or "windows" or "super" or "meta":
                    modifiers |= Win32.MOD_WIN;
                    continue;
            }
            if (vk != 0) return false; // two non-modifier keys: not a hotkey
            if (!KeyNames.TryGetVirtualKey(part, out vk)) return false;
        }
        if (vk == 0) return false;
        binding = new HotkeyBinding(modifiers, vk);
        return true;
    }

    internal string Format()
    {
        if (IsEmpty) return string.Empty;
        var parts = new List<string>(4);
        if ((Modifiers & Win32.MOD_CONTROL) != 0) parts.Add("Ctrl");
        if ((Modifiers & Win32.MOD_ALT) != 0) parts.Add("Alt");
        if ((Modifiers & Win32.MOD_SHIFT) != 0) parts.Add("Shift");
        if ((Modifiers & Win32.MOD_WIN) != 0) parts.Add("Win");
        parts.Add(KeyNames.GetName(VirtualKey));
        return string.Join("+", parts);
    }

    /// <summary>Packs into the LOWORD shape HKM_SETHOTKEY expects.</summary>
    internal ushort ToControlValue()
    {
        int flags = 0;
        if ((Modifiers & Win32.MOD_SHIFT) != 0) flags |= HotkeyfShift;
        if ((Modifiers & Win32.MOD_CONTROL) != 0) flags |= HotkeyfControl;
        if ((Modifiers & Win32.MOD_ALT) != 0) flags |= HotkeyfAlt;
        return (ushort)(((uint)flags << 8) | (VirtualKey & 0xFF));
    }

    /// <summary>
    /// Unpacks an HKM_GETHOTKEY result. The control has no concept of the Windows key, so a
    /// Win-modified binding entered by hand in config.json survives a round trip through the
    /// settings window only if the user does not re-record that field.
    /// </summary>
    internal static HotkeyBinding FromControlValue(ushort value)
    {
        uint vk = (uint)(value & 0xFF);
        int flags = (value >> 8) & 0xFF;
        uint modifiers = 0;
        if ((flags & HotkeyfShift) != 0) modifiers |= Win32.MOD_SHIFT;
        if ((flags & HotkeyfControl) != 0) modifiers |= Win32.MOD_CONTROL;
        if ((flags & HotkeyfAlt) != 0) modifiers |= Win32.MOD_ALT;
        return vk == 0 ? None : new HotkeyBinding(modifiers, vk);
    }
}

internal static class KeyNames
{
    private static readonly Dictionary<string, uint> ByName = Build();
    private static readonly Dictionary<uint, string> ByKey = Invert(ByName);

    internal static bool TryGetVirtualKey(string lowerName, out uint vk) => ByName.TryGetValue(lowerName, out vk);

    internal static string GetName(uint vk) => ByKey.TryGetValue(vk, out string? name) ? name : "0x" + vk.ToString("X2");

    private static Dictionary<string, uint> Build()
    {
        var map = new Dictionary<string, uint>(StringComparer.Ordinal);

        for (uint c = 'A'; c <= 'Z'; c++) map[char.ToLowerInvariant((char)c).ToString()] = c;
        for (uint d = '0'; d <= '9'; d++) map[((char)d).ToString()] = d;
        for (uint f = 1; f <= 24; f++) map["f" + f] = 0x6F + f;
        for (uint n = 0; n <= 9; n++) map["numpad" + n] = 0x60 + n;

        map["space"] = 0x20;
        map["enter"] = 0x0D;
        map["return"] = 0x0D;
        map["tab"] = 0x09;
        map["esc"] = 0x1B;
        map["escape"] = 0x1B;
        map["backspace"] = 0x08;
        map["insert"] = 0x2D;
        map["delete"] = 0x2E;
        map["del"] = 0x2E;
        map["home"] = 0x24;
        map["end"] = 0x23;
        map["pageup"] = 0x21;
        map["pagedown"] = 0x22;
        map["left"] = 0x25;
        map["up"] = 0x26;
        map["right"] = 0x27;
        map["down"] = 0x28;
        map["printscreen"] = 0x2C;
        map["scrolllock"] = 0x91;
        map["pause"] = 0x13;
        map["capslock"] = 0x14;
        map["apps"] = 0x5D;
        map["numlock"] = 0x90;
        map["numpadmultiply"] = 0x6A;
        map["numpadadd"] = 0x6B;
        map["numpadsubtract"] = 0x6D;
        map["numpaddecimal"] = 0x6E;
        map["numpaddivide"] = 0x6F;
        // Named rather than punctuated so nothing collides with the "+" separator.
        map["semicolon"] = 0xBA;
        map["plus"] = 0xBB;
        map["comma"] = 0xBC;
        map["minus"] = 0xBD;
        map["period"] = 0xBE;
        map["slash"] = 0xBF;
        map["backquote"] = 0xC0;
        map["lbracket"] = 0xDB;
        map["backslash"] = 0xDC;
        map["rbracket"] = 0xDD;
        map["quote"] = 0xDE;
        return map;
    }

    private static Dictionary<uint, string> Invert(Dictionary<string, uint> source)
    {
        var result = new Dictionary<uint, string>();
        // Canonical display names, chosen before the aliases so "Enter" wins over "Return".
        string[] canonical =
        [
            "space", "enter", "tab", "esc", "backspace", "insert", "delete", "home", "end",
            "pageup", "pagedown", "left", "up", "right", "down", "printscreen", "scrolllock",
            "pause", "capslock", "apps", "numlock", "numpadmultiply", "numpadadd",
            "numpadsubtract", "numpaddecimal", "numpaddivide", "semicolon", "plus", "comma",
            "minus", "period", "slash", "backquote", "lbracket", "backslash", "rbracket", "quote",
        ];
        foreach (string name in canonical)
        {
            if (source.TryGetValue(name, out uint vk)) result[vk] = Display(name);
        }
        foreach ((string name, uint vk) in source)
        {
            if (!result.ContainsKey(vk)) result[vk] = Display(name);
        }
        return result;
    }

    private static string Display(string lowerName) => lowerName switch
    {
        "pageup" => "PageUp",
        "pagedown" => "PageDown",
        "printscreen" => "PrintScreen",
        "scrolllock" => "ScrollLock",
        "capslock" => "CapsLock",
        "numlock" => "NumLock",
        "numpadmultiply" => "NumpadMultiply",
        "numpadadd" => "NumpadAdd",
        "numpadsubtract" => "NumpadSubtract",
        "numpaddecimal" => "NumpadDecimal",
        "numpaddivide" => "NumpadDivide",
        "lbracket" => "LBracket",
        "rbracket" => "RBracket",
        _ when lowerName.Length > 1 && lowerName[0] == 'f' && char.IsAsciiDigit(lowerName[1]) => lowerName.ToUpperInvariant(),
        _ when lowerName.StartsWith("numpad", StringComparison.Ordinal) => "Numpad" + lowerName[6..],
        _ => char.ToUpperInvariant(lowerName[0]) + lowerName[1..],
    };
}

File: src/TextFixMac/Sources/TextFix/Core/HotkeyBinding.swift
// SPDX-License-Identifier: GPL-3.0-only
import AppKit
import Carbon.HIToolbox

/// A global hotkey, stored in config.json as text ("Cmd+Opt+J", "F13") so it stays hand-editable,
/// and converted here to the shapes RegisterEventHotKey and NSEvent expect.
///
/// Displayed with the usual glyphs - the menu bar shows "⌃⌥J", not "Ctrl+Opt+J" - because that is
/// what every other macOS app shows and what is printed on the user's keycaps.
struct HotkeyBinding: Equatable {
    /// Carbon modifier mask: cmdKey / shiftKey / optionKey / controlKey.
    let modifiers: UInt32

    /// A virtual key code (kVK_*), not a character: the binding survives a keyboard layout change.
    let keyCode: UInt32

    static let none = HotkeyBinding(modifiers: 0, keyCode: UInt32.max)

    var isEmpty: Bool { keyCode == UInt32.max }

    // ---- parsing ----

    static func parse(_ text: String?) -> HotkeyBinding? {
        guard let text, !text.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }

        var modifiers: UInt32 = 0
        var keyCode: UInt32?

        for rawPart in text.split(separator: "+") {
            let part = rawPart.trimmingCharacters(in: .whitespaces).lowercased()
            if part.isEmpty { continue }
            switch part {
            case "cmd", "command", "meta", "super", "win", "windows", "⌘":
                modifiers |= UInt32(cmdKey)
                continue
            case "opt", "option", "alt", "⌥":
                modifiers |= UInt32(optionKey)
                continue
            case "ctrl", "control", "⌃":
                modifiers |= UInt32(controlKey)
                continue
            case "shift", "⇧":
                modifiers |= UInt32(shiftKey)
                continue
            default:
                break
            }
            if keyCode != nil { return nil } // two non-modifier keys: not a hotkey
            guard let code = KeyNames.keyCode(for: part) else { return nil }
            keyCode = code
        }
        guard let keyCode else { return nil }
        return HotkeyBinding(modifiers: modifiers, keyCode: keyCode)
    }

    /// The form written to config.json. Spelled out rather than glyphed so the file stays
    /// something you can type into a text editor.
    func format() -> String {
        if isEmpty { return "" }
        var parts: [String] = []
        if modifiers & UInt32(controlKey) != 0 { parts.append("Ctrl") }
        if modifiers & UInt32(optionKey) != 0 { parts.append("Opt") }
        if modifiers & UInt32(shiftKey) != 0 { parts.append("Shift") }
        if modifiers & UInt32(cmdKey) != 0 { parts.append("Cmd") }
        parts.append(KeyNames.name(for: keyCode))
        return parts.joined(separator: "+")
    }

    /// The form shown on screen: ⌃⌥⇧⌘ in Apple's canonical order, then the key.
    func display() -> String {
        if isEmpty { return "" }
        var glyphs = ""
        if modifiers & UInt32(controlKey) != 0 { glyphs += "⌃" }
        if modifiers & UInt32(optionKey) != 0 { glyphs += "⌥" }
        if modifiers & UInt32(shiftKey) != 0 { glyphs += "⇧" }
        if modifiers & UInt32(cmdKey) != 0 { glyphs += "⌘" }
        return glyphs + KeyNames.displayName(for: keyCode)
    }

    // ---- NSEvent bridge, for the recorder in the settings window ----

    static func from(event: NSEvent) -> HotkeyBinding {
        var modifiers: UInt32 = 0
        let flags = event.modifierFlags
        if flags.contains(.control) { modifiers |= UInt32(controlKey) }
        if flags.contains(.option) { modifiers |= UInt32(optionKey) }
        if flags.contains(.shift) { modifiers |= UInt32(shiftKey) }
        if flags.contains(.command) { modifiers |= UInt32(cmdKey) }
        return HotkeyBinding(modifiers: modifiers, keyCode: UInt32(event.keyCode))
    }

    /// F13 through F20 exist precisely because no keyboard sends them by accident, which makes
    /// them the only keys worth claiming globally without a modifier. They are also what a VIA or
    /// QMK board can be told to send from a spare key.
    var isSafeWithoutModifiers: Bool {
        KeyNames.functionKeys13Through20.contains(keyCode)
    }
}

enum KeyNames {
    /// Written as an array of pairs rather than a literal dictionary so the canonical name for a
    /// key code is simply the first entry that mentions it - the aliases below it are accepted on
    /// input and never used for display.
    private static let table: [(String, Int)] = [
        ("a", kVK_ANSI_A), ("b", kVK_ANSI_B), ("c", kVK_ANSI_C), ("d", kVK_ANSI_D),
        ("e", kVK_ANSI_E), ("f", kVK_ANSI_F), ("g", kVK_ANSI_G), ("h", kVK_ANSI_H),
        ("i", kVK_ANSI_I), ("j", kVK_ANSI_J), ("k", kVK_ANSI_K), ("l", kVK_ANSI_L),
        ("m", kVK_ANSI_M), ("n", kVK_ANSI_N), ("o", kVK_ANSI_O), ("p", kVK_ANSI_P),
        ("q", kVK_ANSI_Q), ("r", kVK_ANSI_R), ("s", kVK_ANSI_S), ("t", kVK_ANSI_T),
        ("u", kVK_ANSI_U), ("v", kVK_ANSI_V), ("w", kVK_ANSI_W), ("x", kVK_ANSI_X),
        ("y", kVK_ANSI_Y), ("z", kVK_ANSI_Z),

        ("0", kVK_ANSI_0), ("1", kVK_ANSI_1), ("2", kVK_ANSI_2), ("3", kVK_ANSI_3),
        ("4", kVK_ANSI_4), ("5", kVK_ANSI_5), ("6", kVK_ANSI_6), ("7", kVK_ANSI_7),
        ("8", kVK_ANSI_8), ("9", kVK_ANSI_9),

        ("f1", kVK_F1), ("f2", kVK_F2), ("f3", kVK_F3), ("f4", kVK_F4), ("f5", kVK_F5),
        ("f6", kVK_F6), ("f7", kVK_F7), ("f8", kVK_F8), ("f9", kVK_F9), ("f10", kVK_F10),
        ("f11", kVK_F11), ("f12", kVK_F12), ("f13", kVK_F13), ("f14", kVK_F14), ("f15", kVK_F15),
        ("f16", kVK_F16), ("f17", kVK_F17), ("f18", kVK_F18), ("f19", kVK_F19), ("f20", kVK_F20),

        ("space", kVK_Space), ("enter", kVK_Return), ("return", kVK_Return),
        ("tab", kVK_Tab), ("esc", kVK_Escape), ("escape", kVK_Escape),
        ("delete", kVK_Delete), ("backspace", kVK_Delete),
        ("forwarddelete", kVK_ForwardDelete), ("del", kVK_ForwardDelete),
        ("home", kVK_Home), ("end", kVK_End),
        ("pageup", kVK_PageUp), ("pagedown", kVK_PageDown),
        ("left", kVK_LeftArrow), ("right", kVK_RightArrow),
        ("up", kVK_UpArrow), ("down", kVK_DownArrow),
        ("help", kVK_Help), ("insert", kVK_Help),

        // Named rather than punctuated so nothing collides with the "+" separator.
        ("semicolon", kVK_ANSI_Semicolon), ("plus", kVK_ANSI_Equal), ("equal", kVK_ANSI_Equal),
        ("comma", kVK_ANSI_Comma), ("minus", kVK_ANSI_Minus), ("period", kVK_ANSI_Period),
        ("slash", kVK_ANSI_Slash), ("backquote", kVK_ANSI_Grave),
        ("lbracket", kVK_ANSI_LeftBracket), ("backslash", kVK_ANSI_Backslash),
        ("rbracket", kVK_ANSI_RightBracket), ("quote", kVK_ANSI_Quote),

        ("numpad0", kVK_ANSI_Keypad0), ("numpad1", kVK_ANSI_Keypad1), ("numpad2", kVK_ANSI_Keypad2),
        ("numpad3", kVK_ANSI_Keypad3), ("numpad4", kVK_ANSI_Keypad4), ("numpad5", kVK_ANSI_Keypad5),
        ("numpad6", kVK_ANSI_Keypad6), ("numpad7", kVK_ANSI_Keypad7), ("numpad8", kVK_ANSI_Keypad8),
        ("numpad9", kVK_ANSI_Keypad9),
        ("numpadmultiply", kVK_ANSI_KeypadMultiply), ("numpadadd", kVK_ANSI_KeypadPlus),
        ("numpadsubtract", kVK_ANSI_KeypadMinus), ("numpaddecimal", kVK_ANSI_KeypadDecimal),
        ("numpaddivide", kVK_ANSI_KeypadDivide), ("numpadenter", kVK_ANSI_KeypadEnter),
    ]

    private static let byName: [String: UInt32] = {
        var map: [String: UInt32] = [:]
        map.reserveCapacity(table.count)
        for (name, code) in table { map[name] = UInt32(code) }
        return map
    }()

    private static let byCode: [UInt32: String] = {
        var map: [UInt32: String] = [:]
        // First mention wins, so "enter" beats "return" and "delete" beats "backspace".
        for (name, code) in table where map[UInt32(code)] == nil { map[UInt32(code)] = name }
        return map
    }()

    static let functionKeys13Through20: Set<UInt32> = [
        UInt32(kVK_F13), UInt32(kVK_F14), UInt32(kVK_F15), UInt32(kVK_F16),
        UInt32(kVK_F17), UInt32(kVK_F18), UInt32(kVK_F19), UInt32(kVK_F20),
    ]

    static func keyCode(for lowerName: String) -> UInt32? { byName[lowerName] }

    /// The name written to config.json.
    static func name(for keyCode: UInt32) -> String {
        guard let raw = byCode[keyCode] else { return "0x" + String(keyCode, radix: 16, uppercase: true) }
        switch raw {
        case "pageup": return "PageUp"
        case "pagedown": return "PageDown"
        case "forwarddelete": return "ForwardDelete"
        case "lbracket": return "LBracket"
        case "rbracket": return "RBracket"
        default:
            if raw.count > 1, raw.hasPrefix("f"), raw.dropFirst().allSatisfy(\.isNumber) {
                return raw.uppercased()
            }
            if raw.hasPrefix("numpad") { return "Numpad" + raw.dropFirst(6).capitalizedFirst }
            return raw.capitalizedFirst
        }
    }

    /// The glyph shown on screen. Arrow keys, Return and friends have keycap symbols that every
    /// Mac user reads faster than the word.
    static func displayName(for keyCode: UInt32) -> String {
        switch Int(keyCode) {
        case kVK_Space: return "Space"
        case kVK_Return, kVK_ANSI_KeypadEnter: return "↩"
        case kVK_Tab: return "⇥"
        case kVK_Escape: return "⎋"
        case kVK_Delete: return "⌫"
        case kVK_ForwardDelete: return "⌦"
        case kVK_LeftArrow: return "←"
        case kVK_RightArrow: return "→"
        case kVK_UpArrow: return "↑"
        case kVK_DownArrow: return "↓"
        case kVK_Home: return "↖"
        case kVK_End: return "↘"
        case kVK_PageUp: return "⇞"
        case kVK_PageDown: return "⇟"
        default: return name(for: keyCode).uppercased()
        }
    }
}

extension StringProtocol {
    var capitalizedFirst: String {
        guard let first else { return String(self) }
        return first.uppercased() + dropFirst()
    }
}
